### PR TITLE
Adding scale error retries

### PR DESCRIPTION
**Pull request by krzysied**
**What this PR does / why we need it**:
ScaleWithRetries will retry all retryable errors, not only conflict error.
ref #63030

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/scale.go [hunk 1]
```diff
@@ -71,11 +71,23 @@ func (pe PreconditionError) Error() string {
 
 // RetryParams encapsulates the retry parameters used by kubectl's scaler.
 type RetryParams struct {
-	Interval, Timeout time.Duration
+	Interval    time.Duration
+	Timeout     time.Duration
+	RetryErrors bool
```
Comment (wojtek-t) [inline pkg/kubectl/scale.go]: I don't think that's the right way of doing that.
This RetryError is something won't be used in production code. So effectively what you're doing here is you're polluting production code with testing code. That's usually bad decision.

This should rather use buliding block from scaler and probably duplicate some logic into test/utils/update.go

In particular, without touching kubectl code at all, in test/utils/update.go you can do sth like:
```
cond := ScaleCondition(...)
// opaque that into RetryErrorCondition.
if err := wait.PollImediate()... {}
return wait.PollImeddiate(.., /*here will be your (probably a bit custom) version of scaleHasDesiredReplicas)

Comment (krzysied) [reply pkg/kubectl/scale.go]: OK, I understand.
I will try different approach then.

File: test/utils/update_resources.go [hunk 2]
```diff
@@ -33,10 +34,25 @@ const (
 	waitRetryTImeout    = 5 * time.Minute
 )
 
+func scaleCondition(scaler kubectl.Scaler, namespace, name string, size uint, gr schema.GroupResource, err *error) wait.ConditionFunc {
+	return func() (bool, error) {
+		waitForScale := kubectl.NewRetryParams(updateRetryInterval, updateRetryTimeout)
+		waitForReplicas := kubectl.NewRetryParams(waitRetryInterval, waitRetryTImeout)
+		if *err = scaler.Scale(namespace, name, size, nil, waitForScale, waitForReplicas, gr); *err != nil {
```
Comment (wojtek-t) [inline test/utils/update_resources.go]: Did you read my comment I added yesterday? Did isn't correct.

This isn't what we want, because of couple reasons:
- in theory it can take much more than we want it to take (in terms of wall time)
- it is not correct - scaler.Scale underneath does two things:
(a) changes size of grouping object (e.g. deployment)
(b) waits until cluster state converges to that state

If we correctly updated object and then had error in second phase, we don't want to do (a) again.
If we timedout on waiting for reconciliation, that's a valid error, which we don't want to retry.

Please take a look into my comment and follow what I suggested there.

File: test/utils/update_resources.go [hunk 3]
```diff
@@ -33,11 +36,47 @@ const (
...
+			err = fmt.Errorf("timed out waiting for %q to be synced", name)
+		}
+	}
+	if err != nil {
 		return fmt.Errorf("Error while scaling %s to %d replicas: %v", name, size, err)
 	}
 	return nil
 }
+
+func scaleHasDesiredReplicas(sClient scaleclient.ScalesGetter, gr schema.GroupResource, resourceName string, namespace string, desiredReplicas int32) wait.ConditionFunc {
```
Comment (wojtek-t) [inline test/utils/update_resources.go]: Let's make scaleHasDesiredReplicas in pkg/kubectl/scale.go public method instead of copying it here.

Comment (krzysied) [reply test/utils/update_resources.go]: done

File: test/utils/update_resources.go [hunk 4]
```diff
@@ -21,7 +21,10 @@ import (
 	"time"
 
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/kubernetes/pkg/kubectl"
+
+	scaleclient "k8s.io/client-go/scale"
```
Comment (wojtek-t) [inline test/utils/update_resources.go]: the convention is that if you you introduce aliases, those are suffixes of the path (there are expections coming from long time ago, but we're trying to change that).

So, if any clientgoscale?
But I think you don't need any alias - you don't seem to have "scale" name anywhere in this file.

File: test/utils/update_resources.go [hunk 5]
```diff
@@ -33,10 +36,31 @@ const (
...
 )
 
-func ScaleResourceWithRetries(scaler kubectl.Scaler, namespace, name string, size uint, gr schema.GroupResource) error {
-	waitForScale := kubectl.NewRetryParams(updateRetryInterval, updateRetryTimeout)
-	waitForReplicas := kubectl.NewRetryParams(waitRetryInterval, waitRetryTImeout)
-	if err := scaler.Scale(namespace, name, size, nil, waitForScale, waitForReplicas, gr); err != nil {
+func RetryErrorCondition(condition wait.ConditionFunc) wait.ConditionFunc {
+	return func() (bool, error) {
+		done, err := condition()
+		if err != nil {
```
Comment (wojtek-t) [inline test/utils/update_resources.go]: We should retry everything. Instead, let's use the IsRetryableAPIError function.

File: test/utils/update_resources.go (hunk 4 above)
Comment (krzysied) [reply test/utils/update_resources.go]: alias removed

File: test/utils/update_resources.go (hunk 5 above)
Comment (krzysied) [reply test/utils/update_resources.go]: done

**Comment by wojtek-t** [comment]
I'm approving this for milestone since:
- this is only test change
- this will help us avoid a group of flakes of our scalability tests (which are release blocking) and have cleaner release signal

@jberkus - FYI ^^

File: pkg/kubectl/scale.go [hunk 6]
```diff
@@ -152,7 +152,7 @@ func (s *genericScaler) Scale(namespace, resourceName string, newSize uint, prec
 		err := wait.PollImmediate(
 			waitForReplicas.Interval,
 			waitForReplicas.Timeout,
-			scaleHasDesiredReplicas(s.scaleNamespacer, gr, resourceName, namespace, int32(newSize)))
+			ScaleHasDesiredReplicas(s.scaleNamespacer, gr, resourceName, namespace, int32(newSize)))
```
Comment (liggitt) [inline pkg/kubectl/scale.go]: let's not export this if we don't have to... it looks like the PR doesn't require it

Comment (wojtek-t) [reply pkg/kubectl/scale.go]: Actually, it is used in test/utils/update_resources.go

File: test/utils/update_resources.go [hunk 7]
```diff
@@ -33,10 +35,34 @@ const (
 	waitRetryTImeout    = 5 * time.Minute
 )
 
-func ScaleResourceWithRetries(scaler kubectl.Scaler, namespace, name string, size uint, gr schema.GroupResource) error {
-	waitForScale := kubectl.NewRetryParams(updateRetryInterval, updateRetryTimeout)
-	waitForReplicas := kubectl.NewRetryParams(waitRetryInterval, waitRetryTImeout)
-	if err := scaler.Scale(namespace, name, size, nil, waitForScale, waitForReplicas, gr); err != nil {
```
[the remark this answers is not in the document]
Comment (shyamjvs) [reply test/utils/update_resources.go]: (sorry I thought I submitted this comment from my phone.. but seems like I didn't)

File: test/utils/update_resources.go [hunk 8]
```diff
@@ -33,10 +35,34 @@ const (
...
+}
+
+func ScaleResourceWithRetries(scalesGetter scale.ScalesGetter, namespace, name string, size uint, gr schema.GroupResource) error {
+	scaler := kubectl.NewScaler(scalesGetter)
+	preconditions := &kubectl.ScalePrecondition{
+		Size:            -1,
+		ResourceVersion: "",
+	}
+	cond := RetryErrorCondition(kubectl.ScaleCondition(scaler, preconditions, namespace, name, size, nil, gr))
+	err := wait.PollImmediate(updateRetryInterval, updateRetryTimeout, cond)
```
Comment (shyamjvs) [inline test/utils/update_resources.go]: How about moving this 'waiting for scaled replicas' logic into separate function in pkg/kubectl/scale.go and then just calling that here?

File: test/utils/update_resources.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (wojtek-t) [reply test/utils/update_resources.go]: I already wrote that comment somewhere before.
There are multiple reasons:
1. we want it to fail after waitRetryTimeout (not potentially after 2x longer time, which could happen in that time)
2. if scaling the object succeeded and we just failed waiting for pods, we should start from the beginning and again scale the object - we already did that.

I asked Krzysiek explicitly to do it this way.

File: test/utils/update_resources.go (hunk 8 above)
Comment (wojtek-t) [reply test/utils/update_resources.go]: Yeah - that sounds fine - sth like "WaitForScaleHasDesiredReplicas(scalesGetter, gr, name, namespace, size, RetryParams)

File: test/utils/update_resources.go [hunk 9]
```diff
@@ -33,10 +35,34 @@ const (
...
+	if err == nil {
+		err = wait.PollImmediate(
+			waitRetryInterval,
+			waitRetryTImeout,
```
Comment (wojtek-t) [inline test/utils/update_resources.go]: Since you're already here:
s/waitRetryTImeout/waitRetryTimeout/

Comment (krzysied) [reply test/utils/update_resources.go]: done

File: test/utils/update_resources.go (hunk 8 above)
Comment (krzysied) [reply test/utils/update_resources.go]: done

File: pkg/kubectl/scale.go [hunk 10]
```diff
@@ -177,3 +170,16 @@ func scaleHasDesiredReplicas(sClient scaleclient.ScalesGetter, gr schema.GroupRe
 			desiredReplicas == actualScale.Status.Replicas, nil
 	}
 }
+
+// WaitForScaleHasDesiredReplicas waits until condition scaleHasDesiredReplicas is satisfied
+// or returns error when timeout happens
+func WaitForScaleHasDesiredReplicas(sClient scaleclient.ScalesGetter, gr schema.GroupResource, resourceName string, namespace string, newSize uint, waitForReplicas *RetryParams) error {
+	err := wait.PollImmediate(
```
Comment (wojtek-t) [inline pkg/kubectl/scale.go]: If waitForReplicas is nil, it will crash.

if waitForReplicas != nil should be handled here.

File: test/utils/update_resources.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (shyamjvs) [reply test/utils/update_resources.go]: I don't think that's true.

> if scaling the object succeeded and we just failed waiting for pods, we should start from the beginning and again scale the object - we already did that.

I've also considered this, and FMU it doesn't matter. Even if we fail in the 'waiting' part and retry the whole operation that's fine, because then the 'scale' would just be a no-op if we try to set it to the same replicas again. See:

```
$ kubectl get rc -n e2e-tests-load-30-nodepods-44-l9njz load-small-9993 -o yaml
...
resourceVersion: "12120458"
replicas: 2

$ kubectl scale -n e2e-tests-load-30-nodepods-44-l9njz --replicas=2 rc/load-small-9993
replicationcontroller/load-small-9993 scaled

$ kubectl get rc -n e2e-tests-load-30-nodepods-44-l9njz load-small-9993 -o yaml
...
resourceVersion: "12120458"
replicas: 2

$ kubectl scale -n e2e-tests-load-30-nodepods-44-l9njz --replicas=2 rc/load-small-9993 --v=8
GET http://localhost:8080/api/v1/namespaces/e2e-tests-load-30-nodepods-44-l9njz/replicationcontrollers/load-small-9993
^ Only GET, no subsequent write call happens
```

> we want it to fail after waitRetryTimeout (not potentially after 2x longer time, which could happen in that time)

We don't wait for 2x longer. We should fail if there was 'any timeout' (either during scaling or waiting), which is correct behaviour IMO. However, if there's some retriable error, we'll redo the whole operation.

In short, I think this PR should be a 2-line change:

```
if err != nil && err != wait.Timeout && IsRetryable(err); then retry; else fail
```

[the remark this answers is not in the document]
Comment (wojtek-t) [reply test/utils/update_resources.go]: > We don't wait for 2x longer. We should fail if there was 'any timeout' (either during scaling or waiting), which is correct behaviour IMO. However, if there's some retriable error, we'll redo the whole operation.

That isn't true. If 'waiting for pods will fail right before timeout on some other error, then:
- that function will return with an error
- the external "wait.Poll" will still be before timeout, so it will trigger the `Scale` one more time
- the new Sclae will start with the original timeouts from scratch

Yes - we can start adjusting timeouts ourselves, but I really don't like this option.

> because then the 'scale' would just be a no-op if we try to set it to the same replicas again. See:

Yes I know it no-op and it won't even hit etcd because we check whether what we are trying to write is exactly the same or not.
But still - that is unnecessary call to apiserver, which I don't want to send.

But the first argument is much more important here.

File: pkg/kubectl/scale.go (hunk 10 above)
Comment (krzysied) [reply pkg/kubectl/scale.go]: done

File: test/utils/update_resources.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (shyamjvs) [reply test/utils/update_resources.go]: I think given that this is just test code for retrying we probably don't need to be so critical with the retry logic. Yes, if we fail in the waiting step, we'll potentially add:
- 1 extra GET call during next scale step (maybe some more if the get fails)
- 1 extra round of waiting for max 5 mins

But given that we shouldn't be needing this retry logic for >99% of the calls likely, the above shouldn't cause much difference imo. Btw - I'm also willing to change waitRetryTimeout to half (i.e 160s) as this is the only purpose for it's use and that should already offer enough retries (= 5) :)

One more thing - the cost of duplicating the whole function (and keeping it in sync with kubectl) is probably more than the benefit I feel.

[the remark this answers is not in the document]
Comment (wojtek-t) [reply test/utils/update_resources.go]: > 1 extra round of waiting for max 5 mins

That changes our expectations - something that was failing previously may pass now - we may hide potential regressions.

> One more thing - the cost of duplicating the whole function (and keeping it in sync with kubectl) is probably more than the benefit I feel.

In this particular case I don't agree - we the current refactoring we pretty much don't have any significnat duplication - there are two building block function that we are reusing  - the glueing code is trivial in this case.

I have a very strong opinion that this approach is better.

[the remark this answers is not in the document]
Comment (shyamjvs) [reply test/utils/update_resources.go]: > That changes our expectations - something that was failing previously may pass now - we may hide potential regressions.

That's why I was suggesting halving the timeout (to keep wait the same). With that we should catch at least those regressions we're catching now.

> there are two building block function that we are reusing - the glueing code is trivial in this case

Yes, that was reason why I had the comment on moving wait to separate function. IMO there's some machinery there we're unnecessarily duplicating (e.g initializing `preconditions`, using some function signatures, etc).

I wouldn't insist, but when I think of it, it seems like we're making a 60-line change instead of a 3-line change just because we don't want to half the timeout (which is only used by this function :)

[the remark this answers is not in the document]
Comment (wojtek-t) [reply test/utils/update_resources.go]: I don't want to half the timeout just to make this PR smaller.
If we believe that half of the timeout is enough, let's change that independently and we that we're back to the same problem. This PR is not supposed to change the default behavior (in case on errors).

[the remark this answers is not in the document]
Comment (shyamjvs) [reply test/utils/update_resources.go]: To begin with, I'm not sure what's so specific with that number (iiuc it's sth that we just informally agreed as a reasonable number).

> This PR is not supposed to change the default behavior (in case on errors).

Who decides what's the default behaviour (besides us)? :)
Just because we've been having that for a while, doesn't mean we can't change it. For e.g if we can catch more regressions by changing it, then we probably should.

Also IIUC from a user perspective they'll call `kubectl scale --replicas=X --wait-for-replicas` and if it fails they'll just call it again. Wdyt?

[the remark this answers is not in the document]
Comment (wojtek-t) [reply test/utils/update_resources.go]: > To begin with, I'm not sure what's so specific with that number (iiuc it's sth that we just informally agreed as a reasonable number).

Nothing specific. But if we agree that X is a reasonable number, the we should use X, not "something random between X and 2X".

> For e.g if we can catch more regressions by changing it, then we probably should.

I agree. But by doing what you suggest we can only catch less regressions, not more.

> Also IIUC from a user perspective they'll call kubectl scale --replicas=X --wait-for-replicas and if it fails they'll just call it again.

If you're a user, then you know how long you want to wait. If you're programming (which we do), we want to have a well defined timeout.

[the remark this answers is not in the document]
Comment (shyamjvs) [reply test/utils/update_resources.go]: > But if we agree that X is a reasonable number, the we should use X, not "something random between X and 2X".

That's not true. Currently we're randomly waiting for "sth b/w 0 and X", with my suggestion it'll be "sth b/w 0 and 2X" :)

> But by doing what you suggest we can only catch less regressions, not more.

Why do you say so? We still continue to retry for max 5 minutes (but split across 2 cycles).

> If you're a user, then you know how long you want to wait. If you're programming (which we do), we want to have a well defined timeout.

My point was not about timeout, but the operation granularity at which we retry. Btw - I think some users would want to exercise timeout (using --timeout flag kubectl exposes), for e.g while doing a large scale-up operation?

[the remark this answers is not in the document]
Comment (wojtek-t) [reply test/utils/update_resources.go]: > That's not true. Currently we're randomly waiting for "sth b/w 0 and X", with my suggestion it'll be "sth b/w 0 and 2X" :)

Well - depends how you're looking into that. Yes - if some error happens, we may end up before X. But there is hard timeout of X.
With your suggestion there won't be hard timeout. Depending what happened in the meantime, the operation may fail *with timeout error* after anything between [X, 2X] - that's pretty bad experience.

> Why do you say so? We still continue to retry for max 5 minutes (but split across 2 cycles).

That isn't true. Imagine a bug that apiserver is crashing every 5 minutes. And imagine that now we regressed and 1% of RCs take more than 5m to scale.
With my change, we will discover it deterministacally. With your, we may or may not discover it.

> My point was not about timeout, but the operation granularity at which we retry. Btw - I think some users would want to exercise timeout (using --timeout flag kubectl exposes), for e.g while doing a large scale-up operation?

Yes, but if it fails after almost timeout (with some random error) - what they do? If they were waiting for 10 minutes, they probably won't wait for another 10. And I don't want to play with timeouts.

[the remark this answers is not in the document]
Comment (shyamjvs) [reply test/utils/update_resources.go]: > With your suggestion there won't be hard timeout

There is a hard timeout, it's 2X. To clarify - with my change, we're just splitting the wait across two iterations (as X + X) instead of waiting in single iteration for 2X.
And halving "waitRetryTimeout" will achieve the same effect.

> With your, we may or may not discover it.

We will detect it with my change. We just need to adjust overall timeout to 5 mins :)

Btw - I'm fine with the current change. I don't mind much that the size of PR is big. But if we can achieve similar effect with a 3-line change, just saying that I'd prefer it more.

[the remark this answers is not in the document]
Comment (wojtek-t) [reply test/utils/update_resources.go]: > There is a hard timeout, it's 2X. To clarify - with my change, we're just splitting the wait across two iterations (as X + X) instead of waiting in single iteration for 2X.
And halving "waitRetryTimeout" will achieve the same effect.

No - there is significantly different UX with your proposal which in my opinion is wrong.
With the current PR:
- if you get "timeout" error" then we know it took X
With your proposal,
- if you get "timeout" error, then it took some time between X and 2X
That is really misleading.

> We will detect it with my change. We just need to adjust overall timeout to 5 mins :)

But that means we need to start playing with timeouts. So when the internal function (Scale) fails with error, we need to call it with a timeout that is original timeout minus time that already passed.
That's something I want to avoid - I prefer some small duplication of code that is fairly trivial.

> But if we can achieve similar effect with a 3-line change, just saying that I'd prefer it more.

What I claim is that we can't achieve the same.

[the remark this answers is not in the document]
Comment (shyamjvs) [reply test/utils/update_resources.go]: That sounds reasonable. Yes, there's some UX difference, but given that it's just retry logic in our test I felt we probably don't need to worry too much about that.

Anyway, let's get this in to stop our flakes.
